### DISCUSSION TITLE
How to access ActionTarget from code-behind

### DISCUSSION THREAD

**Discussion by VisualAlf** [canton7/Stylet#373] [Q&A]
I'd like to have access to viewmodel's methods from code-behind. Something aka {s:Action <method>}.

I know I can use the datacontext, but I'd have to cast that to the name of the VM. (would crash, if name changed for some reason).

As example, I'd like to catch right mouse clicks on some gui element and the VM should display some info on that. Binding the event using eg. {s:Action DoRightMouse} (in xaml) would give the VM access to the MouseButtonsEventArgs, telling about the GUI internal widgets like TextBlocks and so on. The VM is not really interested in the internals of the View, but rather needs some _logical_ information on what to display. Therefore I'd like to call a code-behind method where I could assemble and parameterize all info of interest to the VM. Then, it would be nice, if I could call into the current ActionTarget (wherever that is) to have my info displayed.

Is that possible?

Thxs (also for the excellent software gem, Stylet)
Alfons

**Comment by canton7** [comment]
To be honest, casting the DataContext (or ActionTarget) to the ViewModel's type is *more* safe than using `s:Action`. If the name of the VM changes either VS will update the cast for you, or it will fail to compile because you're casting to a type which no longer exists, whereas `s:Action` will only fail at runtime.

If you want to imitate what `s:Action` does, you can call into the VM using reflection (which is how `s:Action` does it). But there's little point in doing that when you can cast the DataContext to the VM's type, and then get compile-time checking that all of the method invocations on that type are correct.

**Reply by canton7** [reply]
ActionTarget is just an attached property, which you can access with `View.GetActionTarget(view)`

**Reply by VisualAlf** [reply]
I see. 
I just hoped, I could somehow reuse some stuff from s:Action.

Ok. So the KISS way of doing it would be:

in code-behind (ShellView.xaml.cs): 
```
       private void Button_Click(object sender, RoutedEventArgs e)
        {
            var VM = this.DataContext as ShellViewModel;
            if (VM == null) {
                MessageBox.Show("DataContext doesn't match", "ShellView.Button_Click", MessageBoxButton.OK, MessageBoxImage.Error);
                return;
            }
            VM.ExecFromCodeBehind(42, true, "it works");
        }
```
and in ViewModel (ShellViewModel.cs):
```
        public void ExecFromCodeBehind(int i, bool b, string s)
        { 
        }
```
Agreed?

**Reply by canton7** [reply]
I'd just do a normal cast, rather than showing a messagebox with a message which is both incomprehensible and useless to the user. If for some reason you change which VM is associated with a view, you'll pick up the InvalidCastException during your testing.

But otherwise, yes.
